test/Microsoft.AspNetCore.Mvc.Analyzers.Test/Infrastructure/AnalyzerTestBase.cs

```diff
@@ -0,0 +1,131 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Simplification;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.Extensions.DependencyModel;
+
+namespace Microsoft.AspNetCore.Mvc.Analyzers.Infrastructure
+{
+    public abstract class AnalyzerTestBase : IDisposable
+    {
+        public Workspace Workspace { get; private set; }
+
+        protected abstract DiagnosticAnalyzer DiagnosticAnalyzer { get; }
+
+        protected virtual CodeFixProvider CodeFixProvider { get; }
+
+        protected Project CreateProject(string source)
+        {
+            var projectId = ProjectId.CreateNewId(debugName: "TestProject");
+            var newFileName = "Test.cs";
+            var documentId = DocumentId.CreateNewId(projectId, debugName: newFileName);
+            var metadataReferences = DependencyContext.Default.CompileLibraries
+                .SelectMany(c => c.ResolveReferencePaths())
+                .Select(path => MetadataReference.CreateFromFile(path))
+                .Cast<MetadataReference>()
+                .ToList();
+
+            Workspace = Workspace ?? new AdhocWorkspace();
```

Comment:
OH! Btw all creation of an `AdhocWorkspace` needs to be protected by a global (process-level) lock. There is a bug in MEF on desktop framework that we hit recently in razor.

Reply:
Ahh, I saw these errors randomly and I thought I had a test bug (the tests generated by the analyzer template have a lot of Task.Result). Ok, I'll see what you guys did in Razor